File: src/main/java/net/shadowfacts/underwaterutilities/block/tank/BlockTank.java
package net.shadowfacts.underwaterutilities.block.tank;

import net.minecraft.block.material.Material;
import net.minecraft.block.properties.PropertyInteger;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.client.model.ModelLoader;
import net.shadowfacts.underwaterutilities.UnderwaterUtilities;
import net.shadowfacts.underwaterutilities.gui.UUGUIHandler;
import net.shadowfacts.shadowmc.block.BlockTE;
import net.shadowfacts.shadowmc.oxygen.OxygenCaps;
import net.shadowfacts.shadowmc.oxygen.OxygenHandler;
import net.shadowfacts.shadowmc.oxygen.impl.OxygenHandlerImpl;
import net.shadowfacts.shadowmc.util.MeshWrapper;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;

/**
 * @author shadowfacts
 */
public class BlockTank extends BlockTE<TileEntityTank> {

	public static final PropertyInteger LEVEL = PropertyInteger.create("level", 0, 10);

	private static final AxisAlignedBB BOX = new AxisAlignedBB(4/16f, 0, 4/16f, 12/16f, 13/16f, 12/16f);

	public BlockTank() {
		super(Material.ROCK, "oxygenTank");
		setDefaultState(getDefaultState().withProperty(LEVEL, 0));
	}

	@Override
	public void initItemModel() {
		ModelLoader.setCustomMeshDefinition(Item.getItemFromBlock(this), MeshWrapper.of(stack -> {
			OxygenHandler handler = stack.getCapability(OxygenCaps.HANDLER, null);
			int level = (int)((handler.getStored() / (float)handler.getCapacity()) * 10);
			return new ModelResourceLocation(UnderwaterUtilities.modId + ":oxygenTank", "level=" + level);
		}));
	}

	@Nonnull
	@Override
	@Deprecated
	public AxisAlignedBB getBoundingBox(IBlockState state, IBlockAccess source, BlockPos pos) {
		return BOX;
	}

	@Override
	@Deprecated
	public AxisAlignedBB getCollisionBoundingBox(IBlockState blockState, @Nonnull World world, @Nonnull BlockPos pos) {
		return BOX;
	}

	@Nonnull
	@Override
	@Deprecated
	public AxisAlignedBB getSelectedBoundingBox(IBlockState state, @Nonnull World world, @Nonnull BlockPos pos) {
		return BOX;
	}

	@Override
	public boolean isSideSolid(IBlockState state, @Nonnull IBlockAccess world, @Nonnull BlockPos pos, EnumFacing side) {
		return false;
	}

	@Override
	@Deprecated
	public boolean isOpaqueCube(IBlockState state) {
		return false;
	}

	@Override
	public IBlockState onBlockPlaced(World world, BlockPos pos, EnumFacing facing, float hitX, float hitY, float hitZ, int meta, EntityLivingBase placer) {
		return super.onBlockPlaced(world, pos, facing, hitX, hitY, hitZ, meta, placer);
	}

	@Override
	public void onBlockPlacedBy(World world, BlockPos pos, IBlockState state, EntityLivingBase placer, ItemStack stack) {
		TileEntityTank tank = getTileEntity(world, pos);
		tank.loadOxygen(stack.getCapability(OxygenCaps.HANDLER, null));
	}

	@Override
	public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumHand hand, @Nullable ItemStack heldItem, EnumFacing side, float hitX, float hitY, float hitZ) {
		player.openGui(UnderwaterUtilities.instance, UUGUIHandler.TANK, world, pos.getX(), pos.getY(), pos.getZ());
		return true;
	}

	@Nonnull
	@Override
	@Deprecated
	protected BlockStateContainer createBlockState() {
		return new BlockStateContainer(this, LEVEL);
	}

	@Override
	public int getMetaFromState(IBlockState state) {
		return state.getValue(LEVEL);
	}

	@Nonnull
	@Override
	@Deprecated
	public IBlockState getStateFromMeta(int meta) {
		return getDefaultState().withProperty(LEVEL, meta);
	}

	@Nonnull
	@Override
	@Deprecated
	public IBlockState getActualState(@Nonnull IBlockState state, IBlockAccess world, BlockPos pos) {
		return state;
	}

	@Override
	@Deprecated
	public void breakBlock(@Nonnull World world, @Nonnull BlockPos pos, @Nonnull IBlockState state) {
		ItemStack stack = new ItemStack(this);
		((OxygenHandlerImpl)stack.getCapability(OxygenCaps.HANDLER, EnumFacing.NORTH)).load(getTileEntity(world, pos).getCapability(OxygenCaps.HANDLER, EnumFacing.NORTH));

		EntityItem item = new EntityItem(world, pos.getX(), pos.getY(), pos.getZ(), stack);
		world.spawnEntityInWorld(item);

		super.breakBlock(world, pos, state);
	}

	@Nonnull
	@Override
	@Deprecated
	public List<ItemStack> getDrops(IBlockAccess world, BlockPos pos, @Nonnull IBlockState state, int fortune) {
		return new ArrayList<>();
	}

	@Nonnull
	@Override
	public TileEntityTank createTileEntity(@Nonnull World world, @Nonnull IBlockState state) {
		return new TileEntityTank();
	}

	@Override
	public Class<TileEntityTank> getTileEntityClass() {
		return TileEntityTank.class;
	}

}

File: src/main/kotlin/net/shadowfacts/underwaterutilities/block/tank/BlockTank.kt
package net.shadowfacts.underwaterutilities.block.tank

import net.minecraft.block.SoundType
import net.minecraft.block.material.Material

import net.minecraft.block.properties.PropertyInteger
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.minecraftforge.client.model.ModelLoader
import net.shadowfacts.shadowmc.block.BlockTE
import net.shadowfacts.shadowmc.oxygen.OxygenCaps
import net.shadowfacts.shadowmc.oxygen.impl.OxygenHandlerImpl
import net.shadowfacts.shadowmc.util.MeshWrapper
import net.shadowfacts.underwaterutilities.MOD_ID
import net.shadowfacts.underwaterutilities.UnderwaterUtilities
import net.shadowfacts.underwaterutilities.gui.UUGUIHandler
import java.util.ArrayList
/**
 * @author shadowfacts
 */
class BlockTank : BlockTE<TileEntityTank>(Material.ROCK, "oxygen_tank") {

	companion object {
		val LEVEL: PropertyInteger = PropertyInteger.create("level", 0, 10)

		private val BOX = AxisAlignedBB((4 / 16f).toDouble(), 0.0, (4 / 16f).toDouble(), (12 / 16f).toDouble(), (13 / 16f).toDouble(), (12 / 16f).toDouble())
	}

	init {
		defaultState = defaultState.withProperty(LEVEL, 0)
		unlocalizedName = registryName.toString()
		setCreativeTab(CreativeTabs.MISC)
		setHardness(2f)
		soundType = SoundType.STONE
	}

	override fun initItemModel() {
		ModelLoader.setCustomMeshDefinition(Item.getItemFromBlock(this)!!, MeshWrapper.of { stack ->
			val handler = stack.getCapability(OxygenCaps.HANDLER, null)!!
			val level = (handler.stored / handler.capacity.toFloat() * 10).toInt()
			ModelResourceLocation("$MOD_ID:oxygen_tank", "level=" + level)
		})
	}

	@Deprecated("")
	override fun getBoundingBox(state: IBlockState, source: IBlockAccess, pos: BlockPos): AxisAlignedBB {
		return BOX
	}

	override fun getCollisionBoundingBox(blockState: IBlockState, world: IBlockAccess, pos: BlockPos): AxisAlignedBB? {
		return BOX
	}

	@Deprecated("")
	override fun getSelectedBoundingBox(state: IBlockState, world: World, pos: BlockPos): AxisAlignedBB {
		return BOX
	}

	override fun isSideSolid(state: IBlockState, world: IBlockAccess, pos: BlockPos, side: EnumFacing): Boolean {
		return false
	}

	@Deprecated("")
	override fun isOpaqueCube(state: IBlockState): Boolean {
		return false
	}

	@Deprecated("")
	override fun isFullCube(state: IBlockState): Boolean {
		return false
	}

	override fun getStateForPlacement(world: World, pos: BlockPos, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float, meta: Int, placer: EntityLivingBase, hand: EnumHand): IBlockState {
		val stack = placer.getHeldItem(hand)
		val handler = stack.getCapability(OxygenCaps.HANDLER, null)!!
		val level = (handler.stored / handler.capacity * 10).toInt()
		return defaultState.withProperty(LEVEL, level)
	}

	override fun onBlockPlacedBy(world: World, pos: BlockPos, state: IBlockState, placer: EntityLivingBase, stack: ItemStack) {
		val tank = getTileEntity(world, pos)
		tank.loadOxygen(stack.getCapability(OxygenCaps.HANDLER, null)!!)
	}

	override fun onBlockActivated(world: World, pos: BlockPos, state: IBlockState, player: EntityPlayer, hand: EnumHand, side: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): Boolean {
		player.openGui(UnderwaterUtilities, UUGUIHandler.TANK, world, pos.x, pos.y, pos.z)
		return true
	}

	@Deprecated("")
	override fun createBlockState(): BlockStateContainer {
		return BlockStateContainer(this, LEVEL)
	}

	override fun getMetaFromState(state: IBlockState): Int {
		return state.getValue(LEVEL)
	}

	@Deprecated("")
	override fun getStateFromMeta(meta: Int): IBlockState {
		return defaultState.withProperty(LEVEL, meta)
	}

	@Deprecated("")
	override fun breakBlock(world: World, pos: BlockPos, state: IBlockState) {
		val stack = ItemStack(this)
		(stack.getCapability(OxygenCaps.HANDLER, EnumFacing.NORTH) as OxygenHandlerImpl).load(getTileEntity(world, pos).getCapability(OxygenCaps.HANDLER, EnumFacing.NORTH))

		val item = EntityItem(world, pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble(), stack)
		world.spawnEntity(item)

		super.breakBlock(world, pos, state)
	}

	@Deprecated("")
	override fun getDrops(world: IBlockAccess, pos: BlockPos, state: IBlockState, fortune: Int): List<ItemStack> {
		return ArrayList()
	}

	override fun createTileEntity(world: World, state: IBlockState): TileEntityTank {
		return TileEntityTank()
	}

	override fun getTileEntityClass(): Class<TileEntityTank> {
		return TileEntityTank::class.java
	}

}
